### PR TITLE
dispose buffers of C* idle pool members

### PR DESCRIPTION
**Pull request by clockfort**
Keep underlying connection open,
but for idle/returned connections to pool de-allocate
the backing read buffer because we know we already
read it in at a higher level and no longer require it.

(Background for London/remote peeps)
@andy2palantir was having heaps of trouble (pun intended) with a CLI that reads in a lot of data from the DB that only has a 1GB heap.

At first we assumed he was just doing something bad with his own batching (not true, his batching was quite sane), or that his rows were just very large/wide even though he was batching (not true, we examined the Thrift byte[]s returned)

What I found is that the way we do paging is actually screwing him over. Looking at the things using memory on his stack, we found out only a small portion of active memory buffers were actually in use by his active query, but instead he was wasting a lot of memory that idle connections in his client pool still held references to.

It makes sense that thrift can't make this optimization besides me manually doing it, because it can't know I won't try and read from this connection twice and/or again, so previously the buffer was only cleaned up on connection close (TFramedTransport.close() which kills the socket / SSL / etc, which we don't want normally, the point of this pool is to not incur connection open/churn penalty) or when we checkout a connection again from the pool and do a new query with it.

### CODE REVIEW

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolingContainer.java [hunk 1]
```diff
@@ -106,6 +109,11 @@ protected int getPoolSize() {
             if (resource != null) {
                 if (shouldReuse) {
                     log.debug("Returning {} to pool", resource);
+                    TTransport transport = resource.getInputProtocol().getTransport();
+                    if (transport instanceof TMemoryInputTransport) {
+                        log.debug("During {} check-in, cleaned up an iprot of size {}", resource, transport.getBuffer().length);
```
Comment (schlosna) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolingContainer.java]: `getBuffer()` can return null (though possibly not in this case) and then we'd NPE.

Do we care about any other transport types?

Comment (clockfort) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPoolingContainer.java]: TMemInputTransport is the one that the buffers-that-stay-around-after-reads appear as; unlike reads, I assume it's better at self-policing it's own write buffers, but I can look into it.

I doubt the buffer will be null, but no sense not null-checking it, you're right.

**Comment by schlosna** [comment]
:+1:

**Comment by gbrova** [comment]
Could we also add a short description of this fix to the release notes?

**Pull request merged**